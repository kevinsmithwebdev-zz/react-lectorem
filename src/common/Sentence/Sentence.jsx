import React from 'react';
import Lect from '../Lect/Lect';
import { getFrontSpacer } from '../../utils/punctuation';

import './Sentence.css';

export const TIME_CODE_FUDGE = 0.1;

export const getShouldHighlight = (lect, readTime) =>
  lect.start - TIME_CODE_FUDGE <= readTime &&
  readTime <= lect.end + TIME_CODE_FUDGE;

const aggregateSentence = sentence =>
  sentence.lects.reduce((acc, { text: cur }, i) => `${acc}${getFrontSpacer(cur[0])}${cur}`, '');

export const renderLect = ({
  lect, idx, readTime, showModal, isLast, sentenceTranslation, originalSentence,
}) => {
  const shouldHighlight = getShouldHighlight(lect, readTime);
  const key = idx;
  return (
    <Lect
      lect={ lect }
      isFirst={ idx === 0 }
      isLast={ isLast }
      shouldHighlight={ shouldHighlight }
      key={ key }
      readTime={ readTime }
      showModal={ showModal }
      sentenceTranslation={ sentenceTranslation }
      originalSentence={ originalSentence }
    />
  );
};

const Sentence = ({
  sentence, readTime, showModal, configuration: { shouldShowTranslation, isTranslationBelow },
}) => {
  const translationDirectionClassName = isTranslationBelow ? 'SentenceColumn' : 'SentenceRow';

  const containerWrapperClassName = [
    'Sentence',
    shouldShowTranslation ? translationDirectionClassName : '',
  ].join(' ');

  const originalSentence = aggregateSentence(sentence);

  return (
    <span className={ containerWrapperClassName }>
      <span className='SentenceLectsWrapper'>
        {
          sentence.lects.map(
            (lect, idx) => renderLect({
              lect,
              idx,
              readTime,
              showModal,
              isLast: idx === sentence.lects.length - 1,
              sentenceTranslation: sentence.sentenceTranslation,
              originalSentence,
            }),
          )
        }
      </span>
    </span>
  );
};

export default Sentence;
