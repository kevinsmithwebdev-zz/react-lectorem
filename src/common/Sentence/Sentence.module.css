.Sentence {
  /* display: flex; */
}

.SentenceColumn {
  flex-direction: column;
}

.SentenceRow {
  flex-direction: row;
}

.SentenceLectsWrapper {
  flex: 1;
}

.SentenceTranslationWrapper {
  flex: 1;
  font-style: italic;
}

.SentenceTranslationColumn {
  font-size: 18px;
  padding-bottom: 4px;
}

.SentenceTranslationRow {
  font-size: 24px;
}
