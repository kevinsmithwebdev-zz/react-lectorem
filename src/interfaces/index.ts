export interface LectInterface {
  start?: number,
  end?: number,
  text: string,
  translation?: string,
  explanation?: string,
}

export interface SentenceInterface {
  lects: LectInterface[],
  sentenceTranslation?: string,
}

export interface HeaderInterface {
  imagePath?: string,
}

export interface ParagraphInterface {
  header?: HeaderInterface,
  sentences: SentenceInterface[],
}

export interface StoryDataInterface {
  audioPath: string,
  title?: string,
  subtitle?: string,
  paragraphs: ParagraphInterface[],
}

export interface ConfigurationInterface {
  shouldShowTranslation?: boolean,
  isTranslationBelow?: boolean,
  isBySentence?: boolean,
}