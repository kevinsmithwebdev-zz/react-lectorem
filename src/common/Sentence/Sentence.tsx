import React from 'react';
import Lect from '../Lect/Lect';
import { getFrontSpacer } from '../../utils/punctuation';
import './Sentence.css';
import { LectInterface, SentenceInterface, ConfigurationInterface } from '../../interfaces/index';

export const TIME_CODE_FUDGE = 0.1;

interface RenderLectInterface {
  idx: number,
  readTime: number,
  lect: LectInterface,
  showModal: Function,
  isLast: boolean,
  sentenceTranslation: string,
  originalSentence: string,
}

interface SentenceComponentInterface {
  sentence: SentenceInterface,
  readTime: number,
  showModal: Function,
  configuration: ConfigurationInterface,
}

export const getShouldHighlight = (lect: LectInterface, readTime: number): boolean =>
  lect.start - TIME_CODE_FUDGE <= readTime &&
  readTime <= lect.end + TIME_CODE_FUDGE;

export const aggregateSentence = (sentence: SentenceInterface): string =>
  sentence.lects.reduce((acc, { text: cur }, i) => `${acc}${getFrontSpacer(cur[0])}${cur}`, '');

export const renderLect = ({
  lect, idx, readTime, showModal, isLast, sentenceTranslation, originalSentence,
}: RenderLectInterface) => {
  const shouldHighlight = getShouldHighlight(lect, readTime);
  const key = idx;
  return (
    <Lect
      lect={ lect }
      isFirst={ idx === 0 }
      isLast={ isLast }
      shouldHighlight={ shouldHighlight }
      key={ key }
      showModal={ showModal }
      sentenceTranslation={ sentenceTranslation }
      originalSentence={ originalSentence }
    />
  );
};

const Sentence: React.FC<SentenceComponentInterface> = ({
  sentence, readTime, showModal, configuration: { isTranslationShown, isTranslationBelow },
}) => {
  const translationDirectionClassName = isTranslationBelow ? 'SentenceColumn' : 'SentenceRow';

  const containerWrapperClassName = [
    'Sentence',
    isTranslationShown ? translationDirectionClassName : '',
  ].join(' ');

  const originalSentence = aggregateSentence(sentence);

  return (
    <span className={ containerWrapperClassName }>
      <span className='SentenceLectsWrapper'>
        {
          sentence.lects.map(
            (lect, idx) => renderLect({
                lect,
                idx,
                readTime,
                showModal,
                isLast: idx === sentence.lects.length - 1,
                sentenceTranslation: sentence.sentenceTranslation,
                originalSentence,
              }
            )
          )
        }
      </span>
    </span>
  );
};

export default Sentence;
