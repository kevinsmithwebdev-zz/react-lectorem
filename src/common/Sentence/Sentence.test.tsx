import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Sentence, { getShouldHighlight, aggregateSentence, renderLect, TIME_CODE_FUDGE } from './Sentence';
import storyData from '../../fixtureData/story.data';
import { LectInterface } from '../../interfaces/index';

const mockSentence = storyData.paragraphs[0].sentences[0];

describe('Sentence', () => {
  describe('the main component render', () => {
    const baseProps = {
      sentence: mockSentence,
      readTime: 12.34,
      showModal: () => {},
      configuration: {
        isBySentence: false,
        isTranslationShown: false,
        isTranslationBelow: false,
      },
    };

    const renderer = ShallowRenderer.createRenderer();

    describe('with data', () => {
      it('should render and match snapshot', () => {
        renderer.render(<Sentence {...baseProps} />);
        expect(renderer.getRenderOutput()).toMatchSnapshot();
      });
    });

    describe('with data, isTranslationShown', () => {
      it('should render and match snapshot', () => {
        renderer.render(<Sentence { ...baseProps } configuration={ { ...baseProps.configuration, isTranslationShown: true } }/>);
        expect(renderer.getRenderOutput()).toMatchSnapshot();
      });
    });

    describe('with data, isTranslationBelow', () => {
      it('should render and match snapshot', () => {
        renderer.render(<Sentence { ...baseProps } configuration={ { ...baseProps.configuration, isTranslationBelow: true } }/>);
        expect(renderer.getRenderOutput()).toMatchSnapshot();
      });
    });
  });

  describe('getShouldHighlight', () => {
    const lect = {
      text: 'text123',
      start: 1000,
      end: 2000,
    };

    describe('when readTime does not exist', () => {
      const readTime = undefined as unknown as number;
      it('should return false', () => {
        expect(getShouldHighlight(lect, readTime)).toBe(false);
      });
    });

    describe('when lect does not exist', () => {
      const readTime = 123;
      it('should return false', () => {
        expect(getShouldHighlight(undefined as unknown as LectInterface, readTime)).toBe(false);
      });
    });

    describe('when start does not exist', () => {
      const readTime = 123;
      it('should return false', () => {
        expect(getShouldHighlight({ ...lect, start: undefined } as LectInterface, readTime)).toBe(false);
      });
    });

    describe('when end does not exist', () => {
      const readTime = 123;
      it('should return false', () => {
        expect(getShouldHighlight({ ...lect, end: undefined } as LectInterface, readTime)).toBe(false);
      });
    });

    describe('when readTime is well before start', () => {
      const readTime = lect.start - TIME_CODE_FUDGE * 2;
      it('should return false', () => {
        expect(getShouldHighlight(lect, readTime)).toBe(false);
      });
    });

    describe('when readTime is before start but within fudge', () => {
      const readTime = lect.start - TIME_CODE_FUDGE / 2;
      it('should return true', () => {
        expect(getShouldHighlight(lect, readTime)).toBe(true);
      });
    });

    describe('when readTime is well within start and stop', () => {
      const readTime = lect.start + (lect.end - lect.start) / 2;
      it('should return true', () => {
        expect(getShouldHighlight(lect, readTime)).toBe(true);
      });
    });

    describe('when readTime is after start but within fudge', () => {
      const readTime = lect.end + TIME_CODE_FUDGE / 2;
      it('should return true', () => {
        expect(getShouldHighlight(lect, readTime)).toBe(true);
      });
    });

    describe('when readTime is well after end', () => {
      const readTime = lect.end + TIME_CODE_FUDGE * 2;
      it('should return false', () => {
        expect(getShouldHighlight(lect, readTime)).toBe(false);
      });
    });
  });

  describe('aggregateSentence', () => {
    it('should return the correct string', () => {
      const expectedString = ' FrÃ¨re Jacques.';
      expect(aggregateSentence(mockSentence)).toBe(expectedString);
    });
  });

  describe('renderLect', () => {
    describe('if should be highlighted', () => {
      const mockLect = mockSentence.lects[0];
      const readTime = mockLect.start as number + TIME_CODE_FUDGE / 2;
      const functionProps = {
        idx: 12,
        isLast: false,
        lect: mockLect,
        originalSentence: 'original sentence 123',
        readTime,
        sentenceTranslation: 'sentence translation 123',
        showModal: () => {},
      }
      it('should render and match snapshot', () => {
        expect(renderLect(functionProps)).toMatchSnapshot();
      });
    });

    describe('if should not be highlighted', () => {
      const mockLect = mockSentence.lects[0];
      const readTime = mockLect.end as number + TIME_CODE_FUDGE * 2;
      const functionProps = {
        idx: 12,
        isLast: false,
        lect: mockLect,
        originalSentence: 'original sentence 123',
        readTime,
        sentenceTranslation: 'sentence translation 123',
        showModal: () => {},
      }
      it('should render and match snapshot', () => {
        expect(renderLect(functionProps)).toMatchSnapshot();
      });
    });
  });
});
