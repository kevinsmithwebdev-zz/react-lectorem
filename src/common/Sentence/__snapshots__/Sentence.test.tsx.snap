// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sentence renderLect if should be highlighted should render and match snapshot 1`] = `
<Lect
  isFirst={false}
  isLast={false}
  lect={
    Object {
      "end": 5.511,
      "explanation": "\\"Brother\\" in this sense means the title of a monk. \\"Frère\\" is also where we get our word \\"friar\\"",
      "start": 4.424,
      "text": "Frère",
      "translation": "Brother",
    }
  }
  originalSentence="original sentence 123"
  sentenceTranslation="sentence translation 123"
  shouldHighlight={true}
  showModal={[Function]}
/>
`;

exports[`Sentence renderLect if should not be highlighted should render and match snapshot 1`] = `
<Lect
  isFirst={false}
  isLast={false}
  lect={
    Object {
      "end": 5.511,
      "explanation": "\\"Brother\\" in this sense means the title of a monk. \\"Frère\\" is also where we get our word \\"friar\\"",
      "start": 4.424,
      "text": "Frère",
      "translation": "Brother",
    }
  }
  originalSentence="original sentence 123"
  sentenceTranslation="sentence translation 123"
  shouldHighlight={false}
  showModal={[Function]}
/>
`;

exports[`Sentence the main component render with data should render and match snapshot 1`] = `
<span
  className="Sentence "
>
  <span
    className="SentenceLectsWrapper"
  >
    <Lect
      isFirst={true}
      isLast={false}
      lect={
        Object {
          "end": 5.511,
          "explanation": "\\"Brother\\" in this sense means the title of a monk. \\"Frère\\" is also where we get our word \\"friar\\"",
          "start": 4.424,
          "text": "Frère",
          "translation": "Brother",
        }
      }
      originalSentence=" Frère Jacques."
      sentenceTranslation="Brother James."
      shouldHighlight={false}
      showModal={[Function]}
    />
    <Lect
      isFirst={false}
      isLast={false}
      lect={
        Object {
          "end": 6.533,
          "explanation": "In the English version of the song this is rendered as \\"John\\" but a more accurate translation would be \\"James\\" or \\"Jacob\\", from the Latin \\"Iacobus\\".",
          "start": 5.51,
          "text": "Jacques",
          "translation": "James",
        }
      }
      originalSentence=" Frère Jacques."
      sentenceTranslation="Brother James."
      shouldHighlight={false}
      showModal={[Function]}
    />
    <Lect
      isFirst={false}
      isLast={true}
      lect={
        Object {
          "text": ".",
        }
      }
      originalSentence=" Frère Jacques."
      sentenceTranslation="Brother James."
      shouldHighlight={false}
      showModal={[Function]}
    />
  </span>
</span>
`;

exports[`Sentence the main component render with data, isTranslationBelow should render and match snapshot 1`] = `
<span
  className="Sentence "
>
  <span
    className="SentenceLectsWrapper"
  >
    <Lect
      isFirst={true}
      isLast={false}
      lect={
        Object {
          "end": 5.511,
          "explanation": "\\"Brother\\" in this sense means the title of a monk. \\"Frère\\" is also where we get our word \\"friar\\"",
          "start": 4.424,
          "text": "Frère",
          "translation": "Brother",
        }
      }
      originalSentence=" Frère Jacques."
      sentenceTranslation="Brother James."
      shouldHighlight={false}
      showModal={[Function]}
    />
    <Lect
      isFirst={false}
      isLast={false}
      lect={
        Object {
          "end": 6.533,
          "explanation": "In the English version of the song this is rendered as \\"John\\" but a more accurate translation would be \\"James\\" or \\"Jacob\\", from the Latin \\"Iacobus\\".",
          "start": 5.51,
          "text": "Jacques",
          "translation": "James",
        }
      }
      originalSentence=" Frère Jacques."
      sentenceTranslation="Brother James."
      shouldHighlight={false}
      showModal={[Function]}
    />
    <Lect
      isFirst={false}
      isLast={true}
      lect={
        Object {
          "text": ".",
        }
      }
      originalSentence=" Frère Jacques."
      sentenceTranslation="Brother James."
      shouldHighlight={false}
      showModal={[Function]}
    />
  </span>
</span>
`;

exports[`Sentence the main component render with data, shouldShowTranslation should render and match snapshot 1`] = `
<span
  className="Sentence SentenceRow"
>
  <span
    className="SentenceLectsWrapper"
  >
    <Lect
      isFirst={true}
      isLast={false}
      lect={
        Object {
          "end": 5.511,
          "explanation": "\\"Brother\\" in this sense means the title of a monk. \\"Frère\\" is also where we get our word \\"friar\\"",
          "start": 4.424,
          "text": "Frère",
          "translation": "Brother",
        }
      }
      originalSentence=" Frère Jacques."
      sentenceTranslation="Brother James."
      shouldHighlight={false}
      showModal={[Function]}
    />
    <Lect
      isFirst={false}
      isLast={false}
      lect={
        Object {
          "end": 6.533,
          "explanation": "In the English version of the song this is rendered as \\"John\\" but a more accurate translation would be \\"James\\" or \\"Jacob\\", from the Latin \\"Iacobus\\".",
          "start": 5.51,
          "text": "Jacques",
          "translation": "James",
        }
      }
      originalSentence=" Frère Jacques."
      sentenceTranslation="Brother James."
      shouldHighlight={false}
      showModal={[Function]}
    />
    <Lect
      isFirst={false}
      isLast={true}
      lect={
        Object {
          "text": ".",
        }
      }
      originalSentence=" Frère Jacques."
      sentenceTranslation="Brother James."
      shouldHighlight={false}
      showModal={[Function]}
    />
  </span>
</span>
`;
